tools/multisig/multisig.go

```diff
@@ -0,0 +1,235 @@
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/manifest"
+	filTypes "github.com/filecoin-project/lotus/chain/types"
+	"github.com/zondax/fil-parser/actors"
+	"github.com/zondax/fil-parser/parser"
+	"github.com/zondax/fil-parser/parser/helper"
+	"github.com/zondax/fil-parser/tools"
+	"github.com/zondax/fil-parser/types"
+	"go.uber.org/zap"
+)
+
+const (
+	VERSION       = "V1"
+	MultisigStore = "multisig"
+
+	metadataParams     = "Params"
+	metadataReturn     = "Return"
+	metadataTxnIDField = "TxnID"
+	metadataIDField    = "ID"
+	metadataMethod     = "Method"
+	metadataValue      = "Value"
+
+	txStatusOk = "ok"
+)
+
+var proposeTranslateMap = map[string]string{
+	parser.MethodPropose:         parser.MethodPropose,
+	parser.MethodProposeExported: parser.MethodPropose,
+}
+
+var cancelApproveTranslateMap = map[string]string{
+	parser.MethodApprove:         parser.MethodApprove,
+	parser.MethodApproveExported: parser.MethodApprove,
+	parser.MethodCancel:          parser.MethodCancel,
+	parser.MethodCancelExported:  parser.MethodCancel,
+}
+
+type EventGenerator interface {
+	GenerateMultisigEvents(ctx context.Context, transactions []*types.Transaction, tipsetCid string, tipsetKey filTypes.TipSetKey) (*types.MultisigEvents, error)
+}
+
+type eventGenerator struct {
+	helper *helper.Helper
+	logger *zap.Logger
+}
+
+func NewEventGenerator(helper *helper.Helper, logger *zap.Logger) EventGenerator {
+	return &eventGenerator{
+		helper: helper,
+		logger: logger,
+	}
+}
+
+func (eg *eventGenerator) GenerateMultisigEvents(ctx context.Context, transactions []*types.Transaction, tipsetCid string, tipsetKey filTypes.TipSetKey) (*types.MultisigEvents, error) {
+	events := &types.MultisigEvents{
+		Proposals:    []*types.MultisigProposal{},
+		MultisigInfo: []*types.MultisigInfo{},
+	}
+
+	for _, tx := range transactions {
+		if !strings.EqualFold(tx.Status, txStatusOk) {
+			eg.logger.Sugar().Debug("failed tx found, skipping it")
+			continue
+		}
+
+		metadata, err := tools.ParseTxMetadata(tx.TxMetadata)
+		if err != nil {
+			return nil, err
+		}
+
+		isProposalType := isProposalType(tx.TxType)
+		if isProposalType {
+			proposal := eg.createProposal(ctx, tx, metadata, tipsetCid)
+			events.Proposals = append(events.Proposals, proposal)
+		} else {
+			// Only consider transactions where tx.TxTo is a multisig address
+			// because we only take into account those that change the state of the multisig
+
+			addrTo, err := address.NewFromString(tx.TxTo)
+			if err != nil {
+				eg.logger.Sugar().Errorf("could not parse address. Err: %s", err)
+				return nil, err
+			}
+
+			actorName, err := eg.helper.GetActorNameFromAddress(addrTo, int64(tx.Height), tipsetKey)
+			if err != nil {
+				eg.logger.Sugar().Errorf("could not get actor name from address. Err: %s", err)
+				continue
+			}
+			if !strings.EqualFold(actorName, manifest.MultisigKey) {
```

Comment:
Aren't we filtering twice by multisig type? Txs have been filtered on a previous code before this function being called.